lib/cppcheck.cpp

```diff
@@ -590,6 +590,74 @@ void CppCheck::checkSimplifiedTokens(const Tokenizer &tokenizer)
         executeRules("simple", tokenizer);
 }
 
+#ifdef HAVE_RULES
+
+static const char * pcreErrorCodeToString(const int pcreExecRet)
```

Comment:
Maybe some better information can be written. From pcre_exec() documentation (https://www.pcre.org/original/doc/html/pcreapi.html):

 If pcre_exec() fails, it returns a negative number. The following are defined in the header file:

  PCRE_ERROR_NOMATCH        (-1)

The subject string did not match the pattern.

  PCRE_ERROR_NULL           (-2)

Either code or subject was passed as NULL, or ovector was NULL and ovecsize was not zero.

  PCRE_ERROR_BADOPTION      (-3)

An unrecognized bit was set in the options argument.

  PCRE_ERROR_BADMAGIC       (-4)

PCRE stores a 4-byte "magic number" at the start of the compiled code, to catch the case when it is passed a junk pointer and to detect when a pattern that was compiled in an environment of one endianness is run in an environment with the other endianness. This is the error that PCRE gives when the magic number is not present.

  PCRE_ERROR_UNKNOWN_OPCODE (-5)

While running the pattern match, an unknown item was encountered in the compiled pattern. This error could be caused by a bug in PCRE or by overwriting of the compiled pattern.

  PCRE_ERROR_NOMEMORY       (-6)

If a pattern contains back references, but the ovector that is passed to pcre_exec() is not big enough to remember the referenced substrings, PCRE gets a block of memory at the start of matching to use for this purpose. If the call via pcre_malloc() fails, this error is given. The memory is automatically freed at the end of matching.

This error is also given if pcre_stack_malloc() fails in pcre_exec(). This can happen only when PCRE has been compiled with --disable-stack-for-recursion.

  PCRE_ERROR_NOSUBSTRING    (-7)

This error is used by the pcre_copy_substring(), pcre_get_substring(), and pcre_get_substring_list() functions (see below). It is never returned by pcre_exec().

  PCRE_ERROR_MATCHLIMIT     (-8)

The backtracking limit, as specified by the match_limit field in a pcre_extra structure (or defaulted) was reached. See the description above.

  PCRE_ERROR_CALLOUT        (-9)

This error is never generated by pcre_exec() itself. It is provided for use by callout functions that want to yield a distinctive error code. See the pcrecallout documentation for details.

  PCRE_ERROR_BADUTF8        (-10)

A string that contains an invalid UTF-8 byte sequence was passed as a subject, and the PCRE_NO_UTF8_CHECK option was not set. If the size of the output vector (ovecsize) is at least 2, the byte offset to the start of the the invalid UTF-8 character is placed in the first element, and a reason code is placed in the second element. The reason codes are listed in the following section. For backward compatibility, if PCRE_PARTIAL_HARD is set and the problem is a truncated UTF-8 character at the end of the subject (reason codes 1 to 5), PCRE_ERROR_SHORTUTF8 is returned instead of PCRE_ERROR_BADUTF8.

  PCRE_ERROR_BADUTF8_OFFSET (-11)

The UTF-8 byte sequence that was passed as a subject was checked and found to be valid (the PCRE_NO_UTF8_CHECK option was not set), but the value of startoffset did not point to the beginning of a UTF-8 character or the end of the subject.

  PCRE_ERROR_PARTIAL        (-12)

The subject string did not match, but it did match partially. See the pcrepartial documentation for details of partial matching.

  PCRE_ERROR_BADPARTIAL     (-13)

This code is no longer in use. It was formerly returned when the PCRE_PARTIAL option was used with a compiled pattern containing items that were not supported for partial matching. From release 8.00 onwards, there are no restrictions on partial matching.

  PCRE_ERROR_INTERNAL       (-14)

An unexpected internal error has occurred. This error could be caused by a bug in PCRE or by overwriting of the compiled pattern.

  PCRE_ERROR_BADCOUNT       (-15)

This error is given if the value of the ovecsize argument is negative.

  PCRE_ERROR_RECURSIONLIMIT (-21)

The internal recursion limit, as specified by the match_limit_recursion field in a pcre_extra structure (or defaulted) was reached. See the description above.

  PCRE_ERROR_BADNEWLINE     (-23)

An invalid combination of PCRE_NEWLINE_xxx options was given.

  PCRE_ERROR_BADOFFSET      (-24)

The value of startoffset was negative or greater than the length of the subject, that is, the value in length.

  PCRE_ERROR_SHORTUTF8      (-25)

This error is returned instead of PCRE_ERROR_BADUTF8 when the subject string ends with a truncated UTF-8 character and the PCRE_PARTIAL_HARD option is set. Information about the failure is returned as for PCRE_ERROR_BADUTF8. It is in fact sufficient to detect this case, but this special error code for PCRE_PARTIAL_HARD precedes the implementation of returned information; it is retained for backwards compatibility.

  PCRE_ERROR_RECURSELOOP    (-26)

This error is returned when pcre_exec() detects a recursion loop within the pattern. Specifically, it means that either the whole pattern or a subpattern has been called recursively for the second time at the same position in the subject string. Some simple patterns that might do this are detected and faulted at compile time, but more complicated cases, in particular mutual recursions between two different subpatterns, cannot be detected until run time.

  PCRE_ERROR_JIT_STACKLIMIT (-27)

This error is returned when a pattern that was successfully studied using a JIT compile option is being matched, but the memory available for the just-in-time processing stack is not large enough. See the pcrejit documentation for more details.

  PCRE_ERROR_BADMODE        (-28)

This error is given if a pattern that was compiled by the 8-bit library is passed to a 16-bit or 32-bit library function, or vice versa.

  PCRE_ERROR_BADENDIANNESS  (-29)

This error is given if a pattern that was compiled and saved is reloaded on a host with different endianness. The utility function pcre_pattern_to_host_byte_order() can be used to convert such a pattern so that it runs on the new host.

  PCRE_ERROR_JIT_BADOPTION

This error is returned when a pattern that was successfully studied using a JIT compile option is being matched, but the matching mode (partial or complete match) does not correspond to any JIT compilation mode. When the JIT fast path function is used, this error may be also given for invalid options. See the pcrejit documentation for more details.

  PCRE_ERROR_BADLENGTH      (-32)

This error is given if pcre_exec() is called with a negative value for the length argument.

Reply:
Done in https://github.com/danmar/cppcheck/pull/1384/commits/5a8768563d45c71ce7ed8a48fcf9c475531ae837

There are some error ids, where no additional information is provided at https://www.pcre.org/original/doc/html/pcreapi.html

For those errors, where no additional information is available, the error id is printed.

Reply:
To me, it does not make sense to copy&paste the PCRE documentation into our code.

Reply:
What else do you prefer?

Reply:
The previous solution: The ID of the error, so that a user can look it up in the documentation of PCRE. We should not attempt to maintain the documentation of a third-party library.

Reply:
Ok, Daniel should decide. I am fine with both solutions. After a decision is made, i will update this PR accordingly.

Reply:
> The ID of the error, so that a user can look it up in the documentation of PCRE

In my humble opinion it is not trivial to look up this info.

If we for instance write the ID `PCRE_ERROR_PARTIAL` then if you google that then the first couple of matches only talks about pcre_exec options:

    PCRE_PARTIAL ) Return PCRE_ERROR_PARTIAL for a partial
    PCRE_PARTIAL_SOFT ) match if no full matches are found
    PCRE_PARTIAL_HARD Return PCRE_ERROR_PARTIAL for a partial match
if that is found before a full match

Reply:
hmm.. do you have an opinion about deprecating pcre and use C++11 `<regex>` instead?

I more or less want to deprecate rules completely and only have addons. Rules currently have the advantage that they can be executed directly by Cppcheck. Do you see any other advantages?

Reply:
I did never use rules (or addons!), and so I do not care about keeping them in cppcheck.

Anyway, trying to fix issues of third-party projects, such as bad documentation is not the purpose of cppcheck. If we do so, we are in the trap that we have to keep track of changes in the official PCRE versions (although PCRE1 is legacy code that will not receive major changes anymore). And if we consider rules as a non-core feature, I see even less reason to put much effort in providing excessive documentation for it.

Reply:
I have am using rules actively and in case we replace it with C++11 ```<regex>```, backwards compatibility is required. Otherwise all my custom rules are broken.